### PR TITLE
feat(nestjs): add Auth flow and use in-memory database

### PR DESCRIPTION
**Pull request by prestonlimlianjie**
## Context

This PR demonstrates a possible way to set up the Auth controller, service, and database layer using the nest.js framework. It also sets up an in-memory database so that the application can run right out of the box.

## Approach

As discussed in our Starter Kit kick-off session, we wanted the users of `ts-template` to be able to have a fully functional application without additional steps. One of the problems with the current version is `ts-template` is that it requires our users to 1) set up a database, and 2) run database migrations in order to generate the relevant tables.

As such, this PR sets up the database layer in _development_ mode:
- we use an in-memory sqlite database that works right out of the box
- we get Sequelize to automatically sync the models with the database tables by setting `autoLoadModels: true` and `synchronize: true` so there is no need to run migrations

It is extremely important to remind our users that these settings cannot be used for production, and we should also provide a series of steps/code snippets on how to move the codebase to production.

## Notes
Here is a list of features that will be handled in subsequent PRs:
- Create `MailerModule` to send emails https://github.com/opengovsg/ts-template/issues/844
- Add proper tests to test suite https://github.com/opengovsg/ts-template/issues/845
- Add Sequelize migration scaffold https://github.com/opengovsg/ts-template/issues/846

**New dependencies**:

- `@nestjs/sequelize` : ORM plugin for nest.js
- `sequelize`: ORM library
- `sequelize-typescript`: Typescript plugin for sequelize
- `sqlite3`: Database
- `otplib`: OTP library to generate and verify login OTPs

**New dev dependencies**:

- `@types/sequelize` : Types for sequelize

### CODE REVIEW

File: backend/src/auth/auth.service.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+@Injectable()
+export class AuthService {
+  constructor(
+    @InjectModel(User)
+    private readonly userModel: typeof User,
+    private config: ConfigService
+  ) {}
+
+  private totp = totpFactory.clone({
```
Comment (lamkeewei) [inline backend/src/auth/auth.service.ts:17]: Just wondering if we should encapsulate this as a separate class and inject it instead. At test time, we can then more easily inject a mock for this?

File: backend/src/auth/auth.controller.spec.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import { AuthService } from './auth.service'
+import { ConfigModule } from '../config/config.module'
+import { User } from '../database/models'
+
+describe('AuthController', () => {
+  let controller: AuthController
+  const mockModel = {}
+
+  beforeEach(async () => {
+    const module: TestingModule = await Test.createTestingModule({
```
Comment (lamkeewei) [inline backend/src/auth/auth.controller.spec.ts:13]: Perhaps we can do this? Might avoid having to redeclare `providers` and `controllers` during test time by reusing the original module definition.

```ts
const moduleRef = await Test.createTestingModule({
    imports: [AuthModule],
})
  .overrideProvider(getModelToken(User))
  .useValue(mockModel)
  .compile()
```

File: backend/src/auth/auth.controller.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  async generateOtp(
+    @Res() res: Response,
+    @Body() generateOtpDto: GenerateOtpDto
+  ): Promise<void> {
+    try {
+      await this.authService.generateOtp(generateOtpDto)
+      res.status(HttpStatus.OK).json({ message: 'OTP sent' })
+    } catch (error) {
+      Logger.error(error)
+      res
```
Comment (lamkeewei) [inline backend/src/auth/auth.controller.ts:29]: No change required for now, but we can potentially use [exception filters](https://docs.nestjs.com/exception-filters#exception-filters-1) in the future to standardise error response shapes and logging.

**Review by lamkeewei — APPROVED** [review]
lgtm, left a few minor comments!

File: backend/src/auth/auth.service.ts (hunk 1 above)
Comment (prestonlimlianjie) [reply backend/src/auth/auth.service.ts:17]: Makes sense - created an issue https://github.com/opengovsg/ts-template/issues/870

File: backend/src/auth/auth.controller.spec.ts (hunk 2 above)
Comment (prestonlimlianjie) [reply backend/src/auth/auth.controller.spec.ts:13]: I'll merge this in as is for now - will take create an issue to take a closer look at `ModuleRef` (need a bit of time to wrap my head around it)

**Pull request merged**